src/threadutil.cpp

```diff
@@ -0,0 +1,237 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <threadutil.h>
+
+/*  TODO: using thread_local changes the abi in ways that may not play nice
+    when the c++ stdlib is linked dynamically. Disable it until thorough
+    testing has been done. */
+#undef HAVE_THREAD_LOCAL
+
+#ifdef HAVE_SYS_PRCTL_H
+#include <sys/prctl.h> // For prctl, PR_SET_NAME, PR_GET_NAME
+#endif
+
+#if (defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__DragonFly__))
+#include <pthread_np.h>
+#endif
+
+#if defined(HAVE_THREAD_LOCAL)
+#include <atomic>
+#include <thread>
+
+#elif defined(HAVE_PTHREAD)
+#include <atomic>
+#include <pthread.h>
+
+#else
+#include <mutex>
+#include <thread>
+#include <unordered_map>
+#endif
+
+const std::string UNNAMED_THREAD = "<unnamed>";
+
+struct thread_data_type
+{
+    constexpr thread_data_type() = default;
+    thread_data_type(long id, std::string name) : m_id(id), m_name(std::move(name)){}
+
+    long m_id{0};
+    std::string m_name{""};
+};
+
+void thread_util::set_process_name(const char* name)
+{
+#if defined(PR_SET_NAME)
+    // Only the first 15 characters are used (16 - NUL terminator)
+    ::prctl(PR_SET_NAME, name, 0, 0, 0);
+#elif (defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__DragonFly__))
+    pthread_set_name_np(pthread_self(), name);
+#elif defined(MAC_OSX)
+    pthread_setname_np(name);
+#else
+    // Prevent warnings for unused parameters...
+    (void)name;
+#endif
+}
+
+std::string thread_util::get_process_name()
+{
+#if !defined(CAN_READ_PROCESS_NAME)
+    return "";
+#endif
+
+    char threadname_buff[16];
+    char* pthreadname_buff = (char*)(&threadname_buff);
+
+#if defined(PR_GET_NAME)
+    ::prctl(PR_GET_NAME, pthreadname_buff);
+#elif defined(MAC_OSX)
+    pthread_getname_np(pthread_self(), pthreadname_buff, sizeof(threadname_buff));
+#endif
+    return std::string(pthreadname_buff);
+}
+
+static std::string try_set_internal_name_from_process()
+{
+    std::string procname = thread_util::get_process_name();
+    if (procname.size()) thread_util::set_internal_name(procname);
+    return procname;
+}
+
+
+bool thread_util::rename(std::string name)
+{
+    set_process_name(name.c_str());
+    set_internal_name(name);
+}
+
+
+/*
+ * What follows are three separate platform-dependent implementations of
+ * *name and *id parts of the thread_utils interface.
+ *
+ * If we have thread_local, just keep thread ID and name in a thread_local
+ * global.
+ */
+#if defined(HAVE_THREAD_LOCAL)
+
+static thread_local thread_data_type g_thread_data;
+std::string thread_util::get_internal_name()
+{
+    auto name = g_thread_data.m_name;
+
+    if (g_thread_data.m_name.empty()) {
```

Comment:
Why the fallback?

Reply:
There were lots of `"<unnamed>"` occurrences in the logs and so I figured it made sense to fall back to the process name if we hadn't explicitly named the thread (e.g. `bitcoind`, `bitcoin-qt`). Do you think it makes more sense to just explicitly call `rename` on those main threads?

Reply:
I can see concerns with `get_internal_name` being a misnomer. Maybe we should introduce a `GetName` that first tries the internal name and then falls back to the process name?